File: 3- Méthode TABOUS/8Reine.py
#Méthode TABOUS <--- Garder en mémoire une trace des differents minimum qu'on rencontre au cours du parcours des voisins
#Nom : SAHONIN BAYE CARL ELIAS SIMEON
import random
from math import exp

# TableauEntr = [6, 3, 5, 2, 8, 7, 4, 1]
SolIni = []
for k in range(8):
    SolIni.append(random.randint(1,8))

#function fitness
def f(tab):
    cmpt = 0
    #Verification colonne
    for i in range(8):
        for j in range(8):
            if ( (tab[i] == tab[j]) and (i != j) ):
                cmpt += 1
    #Verification Ligne           
    for i in range(8):
        for j in range(8):
            if ( ( abs(tab[i] - tab[j]) == abs(i-j) ) and (i != j) ):
                cmpt += 1
    return cmpt

def voisinage(tab):
    listabou = [] 
    it = 0 # <-- Compteur d'itération
    nbrVois = 5 # <-- Nombre de voisin
    u = tab # u <-- Solution initial
    
    while (f(u) > 0) : 
        it += 1    
        confns = [] # <-- Liste contenant la valeur des conflits respective de ns
        #Voisinage & conflit de chaque voisin
        ns = [ [] for i in range(nbrVois)]
        for i in range(nbrVois):
            ns[i] = list(range(1,9))
            random.shuffle(ns[i])
            confns.append(f(ns[i]))
        #index du conflit le plus faible      
        for i in range(nbrVois):
            m=confns
            m.index(min(confns))
        v = ns[m.index(min(m))]
        
        print('************************* ITERATION ',it,' *************************')
        print(f'Soit un positionnement possible : ', u , ' de conflit : ',f(u))
        print ('On aura pour voisinage : \n')
        print ('Voisin                    |  Conflit')
        print ('--------------------------|------------')
        for i in ns:
            print(i," | ",f(i))
        print()
        if (f(u) <= f(v)): minf = f(u); minp = u; 
        else : minf = f(v); minp = v;
        
        print('le minimum est : ',minp , ' de conflit : ', minf)

        if ( f(v) < f(u) ) :
            listabou.append(v)
            print("==================> CRITERE D'ASPIRATION SATISFAIT \n \n")
            u = v
        else: 
            print("=========> CRITERE D'ASPIRATION NON SATISFAIT")
            print ("==================> RECHERCHE D'UN NOUVEAU MEILLEUR VOISIN \n \n")
        
    
    print('\n \n')
    print ('==========================< LISTE TABOUS >==========================')
    print()
    print ('              Element de la liste Tabous |  Conflit')
    print ('              ---------------------------|------------')
    for i in listabou:
        print("              ",i," | ",f(i))
    print()
    print('========> Solution optimal : ',u, ' dont le conflit est de :', f(u), '\n')

voisinage(SolIni)

File: 3- Méthode TABOUS/8Reine.java

import java.util.ArrayList;

public class Main {

    public static void main(String[] args) {
        Main m = new Main();
        int k = 0;
        ArrayList<String> listeTabou = new ArrayList<String>();
        int[] tableauFinal = m.newRandomTab(8);
        int[] copie = tableauFinal.clone();
        int actualConflit = m.conflit(tableauFinal);
        int aspiration = actualConflit - 1;
        int[][] voisinage = new int[50][8];
        int[] voisinSelectionner = new int[8];
        String footrint = "";
        listeTabou.add(m.arrayToString(tableauFinal));
        while (m.conflit(copie) > 0){
            k++;
            System.out.println("\n\n\n ======================> ETAPE " + k);
            voisinage = m.genererVoisinage(copie);
            voisinSelectionner = m.selectionnerPlusPetitVoisinage(voisinage);
            footrint = m.arrayToString(voisinSelectionner);

            if(listeTabou.contains(footrint)){
                if(!(m.conflit(voisinSelectionner) <= aspiration)){
                    System.out.println("==> CRITERE D'ASPIRATION NON SATISFAIT");
                    System.out.println("==> RECHERCHE DE NOUVEAU MEILLEUR VOISIN");
                    continue;
                }
            }

            aspiration = m.conflit(voisinSelectionner) - 1;
            copie = voisinSelectionner.clone();
            listeTabou.add(footrint);
            System.out.println("======================================>");
        }
        for(String i: listeTabou){
            System.out.println(i);
        }
        m.printTab(copie, "", "");
    }

    public int[][] genererVoisinage(int[] tab){
        printTab(tab, "====> RECHERCHE DES VOISINS DU TABLEAU", "===> VOISINS TROUVES : ");
        int i = 50;
        int positionToModify;
        int[] copie = new int[8];
        int[][] voisinage = new int[50][8];
        while (i > 0){
            copie = tab.clone();
            positionToModify = (int) ((Math.random() * (7)) + 0);
            copie[positionToModify] = (copie[positionToModify] + (int) ((Math.random() * (10000 - 500)) + 500)) % 8;
            if(eviterDuplication(voisinage, copie)){
                continue;
            }
            voisinage[50 - i] = copie.clone();
            printTab(copie.clone(), "", "");
            i--;
        }
        System.out.println("====> FIN DE LA GENERATION DES VOISINS");
        return voisinage;
    }

    public int[] selectionnerPlusPetitVoisinage(int[][] voisinage){
        System.out.println("====> SELECTION DU VOISIN");
        int conflitPrecedent;
        int[] voisinSelectionner = voisinage[0].clone();
        int conflitActuel = conflit(voisinSelectionner);
        for (int i = 1; i< voisinage.length; i++){
            conflitPrecedent = conflit(voisinage[i]);
            if(conflitPrecedent < conflitActuel){
                voisinSelectionner = voisinage[i].clone();
                conflitActuel = conflit(voisinSelectionner);
            }
        }
        printTab(voisinSelectionner, "Le voisin sélectionné est : ", "====> FIN DE LA SELECTION DES VOISINS");
        return voisinSelectionner;
    }

    public int[] newRandomTab(int size){
        int[] tabToReturn = new int[size];
        for(int i = 0; i < size; i++){
            tabToReturn[i] = (int) ((Math.random() * (7 - 0)) + 0);
        }
        return tabToReturn;
    }

    public int conflit(int[] tab){
        int conflits = 0;
        for(int i  = 0; i < tab.length; i++ ){
            for(int j = 0; j < tab.length; j++){
                conflits += (tab[i] == tab[j]) && (i!=j) ? 1 : 0;
                conflits += (Math.abs(tab[i] - tab[j]) == Math.abs(i - j)) && (i!=j) ? 1 : 0;
            }
        }
        return conflits;
    }

    public String arrayToString(int[] tab){
        StringBuilder str = new StringBuilder();
        for (int j : tab) {
            str.append(j);
        }
        return str.toString();
    }

    public boolean eviterDuplication(int[][] arrayToCheck, int[] arrayToFind){
        boolean found =  false;
        for (int i = 0; i< arrayToCheck.length; i++){
            for (int j = 0; j < arrayToCheck[i].length; j++){
                if(arrayToCheck[i][j] != arrayToFind[j]){
                    found = false;
                    break;
                }
                found = true;
            }
            if(found){
                break;
            }
        }
        return found;
    }

    public void printTab(int[] tab, String startMessage, String endMessage){
        if(!startMessage.equals("")){
            System.out.println(startMessage);
        }

        System.out.print("(");
        for (int i = 0; i < tab.length; i++){
            System.out.print(tab[i]);
            if(i < tab.length - 1){
                System.out.print(",");
            }
        }
        System.out.println(") -- Conflit = " + conflit(tab));

        if(!endMessage.equals("")){
            System.out.println(endMessage);
        }

    }
}
